### ISSUE TITLE
Another false positive `ParadoxicalCondition` when using `in_array()`

### ISSUE THREAD

**Issue by corphi** [vimeo/psalm#6439]
As a follow-up to #6317:

The issue persists when types are used. This snippet triggers the same false positive:

```php
/**
 * @param string[] $list1
 * @param string[] $list2
 */
function contains(array $list1, array $list2, string $element): void
{
    if (in_array($element, $list1, true)) {
    } elseif (in_array($element, $list2, true)) {
    }
}
```

When any of the array types are dropped, no error is reported.

See https://psalm.dev/r/1fc70cf120

**Comment by orklah** [comment]
@TysonAndre that's one thing I was afraid of, and I think this was triggered by adding support for non-literal types in your PR. Is there a specific code that will stop to work correctly without this support?

**Comment by TysonAndre** [comment]
I think my change only affected the sealed array (TKeyedArray, i.e. array literals) and not the branch being referred to, but I may be mistaken

But my change still introduced a bug. For now, the conservative approach of returning the empty array for non-literals would avoid the false positives at the cost of not inferring types.

```php
function contains2(string $a, string $b, $element): void
{
    if (in_array($element, [$a], true)) {
    } elseif (in_array($element, [$b], true)) { // false positive Condition ($element is string) contradicts a previously-established condition ($element is not string)
    }
}
```

The code for handling strict equality in AssertionFinder::scrapeEqualityAssertions may be of use.

Unfortunately,
- The example only works for union types with a single type (literal or non-literal)
- The isSingle check seems to be related to union types only working for getAssertionString with a single type
- `$var_type->getAssertionString() !== $assertion` for atomic types seems useful in avoiding redundant condition warnings, but not paradoxical condition warnings. Type\Union::getAssertionString won't work for variables with more than one type, it only works for union types of a single type, which is why I see isSingle checks elsewhere
- Not sure why that code in scrapeEqualityAssertions is using `~` for approximate equality. Possibly a bug or workaround. Seems to be used in AssertionReconciler::reconcile(). Probably, it's so that negated the condition such as `if ($var === $otherString)` in the else branch isn't treated as a non-string, which happened when I tried to change it.

  Changing it would do a better job of removing mixed, though
- Not sure why getAssertionString() sometimes needs to be prefixed with `=` (strict equality) to be a valid assertion string. I'm guessing there might be special cases for `null` that aren't written for `=null`.

```php
                // both side of the Identical can be asserted to the intersection of both
                $intersection_type = Type::intersectUnionTypes($var_type, $other_type, $codebase);

                if ($intersection_type !== null && $intersection_type->isSingle()) {
                    $assertion = $intersection_type->getAssertionString();

                    $if_types = [];

                    $var_name_left = ExpressionIdentifier::getArrayVarId(
                        $conditional->left,
                        $this_class_name,
                        $source
                    );

                    if ($var_name_left &&
                        (!$var_type->isSingle() || $var_type->getAssertionString() !== $assertion)) {
                        $if_types[$var_name_left] = [['~'.$assertion]];
                    }

                    $var_name_right = ExpressionIdentifier::getArrayVarId(
                        $conditional->right,
                        $this_class_name,
                        $source
                    );

                    if ($var_name_right &&
                        (!$other_type->isSingle() || $other_type->getAssertionString() !== $assertion)) {
                        $if_types[$var_name_right] = [['~'.$assertion]];
                    }

                    return $if_types ? [$if_types] : [];
                }
```

**Comment by TysonAndre** [comment]
```patch
--- a/src/Psalm/Internal/Algebra.php
+++ b/src/Psalm/Internal/Algebra.php
@@ -68,6 +68,10 @@ class Algebra
         if ($type === 'mixed') {
             return $type;
         }
+        if (substr($type, 0, 9) === 'in-array-') {
+            // TODO: Return negation of the subset of literal types instead?
+            return 'mixed';
+        }
 
         return $type[0] === '!' ? substr($type, 1) : '!' . $type;
     }
```

This approach sort of works for avoiding the original false positive for categories with multiple values such as `string`. I assume Type::parseType + filtering out single types such as TFalse would improve accuracy even more

**Comment by orklah** [comment]
Honestly, I'm kinda lost, I only debugged enough to find the biggest flaws in the previous implementation so I don't have all the facts in mind. Care to add this in a PR along with the example in the tests? We'll improve incrementally if needed...

**Comment by TysonAndre** [comment]
Now that I think about it, it should always be `mixed` in the negation. `if (!in_array($element, $array, true))` doesn't give any additional type information when $array can be the empty array.

**Comment by corphi** [comment]
Fixed by #6445. Thanks, @TysonAndre!
